validator/index.js

```diff
@@ -25,50 +25,256 @@ const path = require('path');
 const program = require('commander');
 const vm = require('vm');
 const url = require('url');
+const util = require('util');
 
 /**
- * If the provided fileOrUrl start with 'http://' or 'https://', downloads
- * with the http or https module, otherwise reads the file with the file module.
- * Or, if the provided fileOrUrl is '-', reads from stdin.
- * Then, calls onSuccess with the data, or onFailure with the error message.
- * This method will check the HTTP status code as well and generate
- * brief error messages if it's not 200.
- * @param {!string} fileOrUrl
- * @param {!function(string)} onSuccess
- * @param {!function(string)} onFailure
+ * Convenience function to detect whether an argument is a URL. If not,
+ * it may be a local file.
+ * @param {!string} url
+ * @returns {!boolean}
  */
-function readFileOrDownload(fileOrUrl, onSuccess, onFailure) {
-  const digestChunksFrom = function(response) {
+function isHttpOrHttpsUrl(url) {
+  return url.startsWith('http://') || url.startsWith('https://');
+}
+
+/**
+ * Creates a promise which reads from a file.
+ * @param {!string} name
+ * @returns {!Promise<!string>}
+ */
+function readFromFile(name) {
+  return new Promise(function(resolve, reject) {
+    fs.readFile(name, 'utf8', (err, data) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(data);
+      }
+    });
+  });
+}
+
+/**
+ * Creates a promise which reads from a stream.
+ * @param {!stream.Readable} readable
+ * @returns {!Promise<!string>}
+ */
+function readFromReadable(readable) {
+  return new Promise(function(resolve, reject) {
     const chunks = [];
-    response.setEncoding('utf8');
-    response.on('data', function(chunk) { chunks.push(chunk); });
-    response.on('end', function() { onSuccess(chunks.join('')); });
-  };
-  if (fileOrUrl === '-') {
-    digestChunksFrom(process.stdin);
+    readable.setEncoding('utf8');
+    readable.on('data', (chunk) => { chunks.push(chunk); });
+    readable.on('end', () => { resolve(chunks.join('')); });
+    readable.on('error', (error) => { reject(error); });
+  });
+}
+
+/**
+ * Creates a promise which reads from standard input. Even though it would
+ * be easy to make a function that just returns the data, we return a promise
+ * for consistency with readFromUrl and readFromFile.
+ * @returns {!Promise<!string>}
+ */
+function readFromStdin() {
+  return readFromReadable(process.stdin).then((data) => {
     process.stdin.resume();
-    return;
+    return data;
+  });
+}
+
+/**
+ * Creates a promise which reads from a URL or more precisely, fetches
+ * the contents located at the URL by using the 'http' or 'https' module.
+ * Any HTTP status other than 200 is interpreted as an error.
+ * @param {!string} url
+ * @returns {!Promise<!string>}
+ */
+function readFromUrl(url) {
+  return new Promise(function(resolve, reject) {
+           const clientModule = url.startsWith('http://') ? http : https;
```

Comment:
I tried putting the new on the next line or the function keyword, but the clang formatter keeps putting it back. I think it's trying to line stuff up with 'new' here. :-/